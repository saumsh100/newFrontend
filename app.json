  {
  "name": "carecru",
  "scripts": {
  },
  "env": {
    "NODE_ENV": {
      "required": true
    },
    "RETHINKDB_AUTHKEY": {
      "required": true
    },
    "RETHINKDB_DB": {
      "required": true
    },
    "RETHINKDB_HOST": {
      "required": true
    },
    "RETHINKDB_PORT": {
      "required": true
    },
    "VENDASTA_API_KEY": {
      "required": true
    },
    "VENDASTA_API_USER": {
      "required": true
    },
    "TWILIO_ACCOUNT_SID": {
      "required": true
    },
    "TWILIO_AUTH_TOKEN": {
      "required": true
    },
    "TWILIO_NUMBER": {
      "required": true
    }
  },
  "formation": {
    "web": {
      "quantity": 1
    }
  },
  "addons": [

  ],
  "buildpacks": [
    {
      "url": "heroku/nodejs"
    }
  ]
}
