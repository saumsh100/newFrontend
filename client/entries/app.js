
import '../apps/default';
import '../apps/Dashboard';
