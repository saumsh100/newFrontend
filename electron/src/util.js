
module.exports.getChannelFromVersion = (version) => {
  const splitVersion = version.split('-');
  return splitVersion[1] || null;
};
