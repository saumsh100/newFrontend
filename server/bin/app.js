
require('../config/initializeCodeTranspiler');

const globals = require('../config/globals');
const applyWebpack = require('../config/webpack/applyWebpack');
const handleErrors = require('../middleware/handleErrors');
const chalk = require('chalk');
const app = require('../config/express');

// require('../config/kue');

// Uses the NODE_ENV to determine logging full stack or not
app.set('showStackError', true);

// Load in webpack configurations
applyWebpack(app);


// Set the Handlebars templating engine
app.set('views', `${globals.root}/views`);
app.engine('hbs', require('../config/handlebars').engine);
app.set('view engine', 'hbs');

// Extra logging for communication with server
app.use((req, res, next) => {
  console.log(chalk.blue(req.method, req.originalUrl));
  next();
});


// Bind the routes

app.use(require('../routes'));

// Catch errors, log and respond to client
app.use(handleErrors);

// assume 404 since no middleware responded
/*app.use((req, res) => {
  res.status(404).render('404', {
    url: req.originalUrl,
    error: 'Not found',
  });
});*/

module.exports = app;
