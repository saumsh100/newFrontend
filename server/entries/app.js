
import '../../client/apps/default';
import '../../client/apps/Dashboard';
