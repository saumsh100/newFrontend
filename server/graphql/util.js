
import { getOffsetWithDefault, offsetToCursor } from 'graphql-relay';

/**
 * A simple function that accepts an array and connection arguments, and returns
 * a connection object for use in GraphQL
 *
 * This is a fork of connectionFromArraySlice
 * from https://github.com/graphql/graphql-relay-js/blob/master/src/connection/arrayconnection.js#L65
 *
 * Diferente form the original, is this case we build the cursor and pageInfo based on args
 * whitout slicing the array.
 * That way we can pass the already sliced or filtered array to the function.
 *
 * For now, this only works with forward pagination.
 */
// eslint-disable-next-line import/prefer-default-export
export function connectionFromArrayWithoutSlice(array, args, meta) {
  const { after, first, last } = args;
  const { arrayLength } = meta;

  if (typeof first === 'number') {
    if (first < 0) {
      throw new Error('Argument "first" must be a non-negative integer');
    }
  }
  if (typeof last === 'number') {
    if (last < 0) {
      throw new Error('Argument "last" must be a non-negative integer');
    }
  }

  const afterOffset = getOffsetWithDefault(after, 0);

  const startOffset = Math.max(afterOffset, -1) + 1;
  const endOffset = startOffset + first;

  const edges = array.map((value, index) => ({
    cursor: offsetToCursor(startOffset + index),
    node: value,
  }));

  const firstEdge = edges[0];
  const lastEdge = edges[edges.length - 1];
  const lowerBound = after ? afterOffset + 1 : 0;
  const upperBound = arrayLength;

  return {
    edges,
    pageInfo: {
      startCursor: firstEdge ? firstEdge.cursor : null,
      endCursor: lastEdge ? lastEdge.cursor : null,
      hasPreviousPage: typeof last === 'number' ? startOffset > lowerBound : false,
      hasNextPage: typeof first === 'number' ? endOffset < upperBound : false,
    },
  };
}
