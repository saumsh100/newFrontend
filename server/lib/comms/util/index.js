
import groupBy from 'lodash/groupBy';
import map from 'lodash/map';
import toArray from 'lodash/toArray';

const ATTRS = {
  email: 'email',
  sms: 'mobilePhoneNumber',
  phone: 'mobilePhoneNumber',
};

const ERROR_CODES = {
  email: '1100',
  sms: '1200',
  phone: '1300',
};

const PREFERENCE_TYPES = {
  email: 'emailNotifications',
  sms: 'sms',
  phone: 'phone',
};

const NO_PREFERENCE_ERROR_CODES = {
  email: '2100',
  sms: '2200',
  phone: '2300',
};

/**
 * cannotSend
 *
 * @param patient
 * @param primaryType
 * @returns {{errors: Array, success: Array}}
 */
export function cannotSend(patient, primaryType) {
  // If it is undefined return the error code
  if (!patient[ATTRS[primaryType]]) {
    return ERROR_CODES[primaryType];
  }
}

/**
 * noPreference
 *
 * @param patient
 * @param primaryType
 * @returns {{errors: Array, success: Array}}
 */
export function noPreference(patient, primaryType) {
  const preferences = patient.preferences || {};

  // If it is false return errorCode
  if (!preferences[PREFERENCE_TYPES[primaryType]]) {
    return NO_PREFERENCE_ERROR_CODES[primaryType];
  }
}

/**
 * generateOrganizedPatients will accept an array of patients and an
 * array of primaryTypes and will return the errors and successes based
 * on patientData. The point is to organize the failed responses so that
 * certain jobs can save a bulk amount to save time
 *
 * @param patients
 * @param primaryTypes (ie.// ['email', 'sms'])
 * @returns {{errors: Array, success: Array}}
 */
export function generateOrganizedPatients(patients, primaryTypes) {
  const errors = [];
  const success = [];

  for (const patient of patients) {
    primaryTypes.forEach((primaryType) => {
      const errorCode = cannotSend(patient, primaryType) || noPreference(patient, primaryType);
      if (errorCode) {
        errors.push({
          patient,
          errorCode,
          primaryType,
        });
      } else {
        success.push({ primaryType, patient });
      }
    });
  }

  return { errors, success };
}

/**
 * organizeForOutbox is a function that accepts an array and a selectorPredicate
 * and will group the array based on the selectorPredicate, and then map over that array
 * to return outbox items that have primaryTypes grouped together.
 *
 * NOTE: this is primarily a layer ontop of the mapPatientsTo____ calls that will take the flat
 * array with primaryType and bring it back into primaryTypes
 *
 * @param array (ie.// [{ patient, appointment, primaryType: 'sms' }, { patient, appointment, primaryType: 'email' }])
 * @param selectorPredicate (ie.// ({ patient, appointment }) => `${patient.id}_${appointment.id}`)
 * @param mergePredicate
 * @returns [organizedArray] (ie.// [{ patient, appointment, primaryTypes: ['sms', 'email'] }])
 */
export function organizeForOutbox(array, selectorPredicate, mergePredicate) {
  const groupedMap = groupBy(array, selectorPredicate);
  return toArray(map(groupedMap, mergePredicate));
}
