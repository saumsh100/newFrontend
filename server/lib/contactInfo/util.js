
const orderBy = require('lodash/orderBy');

/**
 * isHead checks if a given patient is the head of the given family by comparing accountId,
 * familyId and headId.
 * @param family
 * @param patient
 * @return {boolean}
 */
const isHead = (family, patient) =>
  !!family &&
  !!patient &&
  family.accountId === patient.accountId &&
  family.id === patient.familyId &&
  family.headId === patient.id;

/**
 * isActive checks if a patient has status equals to the string 'Active'
 * @param patient
 * @return {boolean}
 */
const isActive = ({ status }) => status === 'Active';

/**
 * getFromNewestCreatedFamily orders the a list of patients by its family creation date and
 * grabs the first.
 * @param patients
 */
const getFromNewestCreatedFamily = patients =>
  orderBy(
    patients,
    ({ family: { pmsCreatedAt, createdAt } }) => pmsCreatedAt || createdAt,
    'desc',
  )[0];

/**
 * getOldest orders a list of patients by their birth dates and grabs the first, thus the oldest.
 * @param patients
 */
const getOldest = patients => orderBy(patients, p => p.birthDate, 'asc')[0];

/**
 * getPatientAge is a utility function to extract the patient age form its birth date by year
 * diffing. This was created using plain JS so we can reduce the usage of moment.js.
 * @param patient
 * @return {number}
 */
const getPatientAge = (patient) => {
  const ageDate = new Date(Date.now() - new Date(patient.birthDate).getTime());
  return Math.abs(ageDate.getUTCFullYear() - 1970);
};

/**
 * adultChecker is a configurable function builder that creates a function that will check if an age
 * of a patient is within the adult range provided to it.
 * @param min
 * @param max
 * @return {function(*=): boolean}
 */
const adultChecker = (min, max) => patient =>
  min < getPatientAge(patient) && getPatientAge(patient) < max;

/**
 * grabPoC is a function that takes a patient list and returns a function that filters this list by
 * the given predicate, in case of the filter returns more than one result the fallbackHandler is
 * applied to the list.
 * This function use partially applied arguments so we can reuse the same list of patients but with
 * different predicates and fallbackHandlers.
 * @param patients
 * @return {Function}
 */
const grabPoC = patients => (filterPredicate, fallbackHandler) => {
  const filteredPatients = patients.filter(filterPredicate);
  if (filteredPatients.length === 1) {
    return filteredPatients[0];
  } else if (filteredPatients.length > 1) {
    return fallbackHandler(filteredPatients);
  }

  return undefined;
};

module.exports = {
  adultChecker,
  grabPoC,
  isHead,
  isActive,
  getFromNewestCreatedFamily,
  getOldest,
};
