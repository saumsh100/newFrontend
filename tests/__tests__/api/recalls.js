
import request from 'supertest';
import app from '../../../server/bin/app';
import {
  Account,
  Recall,
} from '../../../server/models';
import wipeModel, { wipeAllModels } from '../../util/wipeModel';
import seedTestUsers from '../../util/seedTestUsers';
import generateToken from '../../util/generateToken';
import { getModelsArray } from '../../util/selectors';

const rootUrl = '/api/accounts';
const accountId = '62954241-3652-4792-bae5-5bfed53d37b7';
const accountId2 = '52954241-3652-4792-bae5-5bfed53d37b7';
const enterpriseId = 'c5ab9bc0-f0e6-4538-99ae-2fe7f920abf4';
const recallId1 = 'd5ab9bc0-f0e6-4538-99ae-2fe7f920abf4';
const recallId2 = 'e5ab9bc0-f0e6-4538-99ae-2fe7f920abf4';
const newRecallId = 'f5ab9bc0-f0e6-4538-99ae-2fe7f920abf4';

async function seedRecalls() {
  await wipeModel(Recall);

  // seed recalls
  await Recall.save([
    {
      id: recallId1,
      accountId,
      primaryType: 'sms',
      createdAt: '2017-07-19T00:14:30.932Z',
    },
    {
      id: recallId2,
      accountId,
      primaryType: 'sms',
      createdAt: '2017-07-19T00:14:30.932Z',
    },
  ]);
}

describe('/api/accounts/:account/recalls', () => {
  // Seed with some standard user data
  let token = null;
  beforeAll(async () => {
    await seedTestUsers();

    // Seed an extra account for fetching multiple and testing switching
    await Account.save({
      id: accountId2,
      enterpriseId,
      name: 'Test Account 2',
      createdAt: '2017-07-20T00:14:30.932Z',
    });

    token = await generateToken({ username: 'manager@test.com', password: '!@CityOfBudaTest#$' });
  });

  afterAll(async () => {
    await wipeAllModels();
  });

  describe('Recalls', () => {
    beforeAll(async () => {
      await seedRecalls();
    });

    describe('GET /:accountId/recalls', () => {
      test('should fetch all recalls for the account', () => {
        return request(app)
          .get(`${rootUrl}/${accountId}/recalls`)
          .set('Authorization', `Bearer ${token}`)
          .expect(200)
          .then(({ body }) => {
            const recalls = getModelsArray('recalls', body);
            expect(recalls.length).toBe(2);
            expect(body).toMatchSnapshot();
          });
      });

      test('should return 403 for different accountIds', () => {
        return request(app)
          .get(`${rootUrl}/${accountId2}/recalls`)
          .set('Authorization', `Bearer ${token}`)
          .expect(403);
      });
    });

    describe('POST /:accountId/recalls', () => {
      test('should create a recall for the account', () => {
        return request(app)
          .post(`${rootUrl}/${accountId}/recalls`)
          .set('Authorization', `Bearer ${token}`)
          .send({
            id: newRecallId,
            primaryType: 'sms',
            createdAt: '2017-07-19T00:14:30.932Z',
          })
          .expect(200)
          .then(({ body }) => {
            const recalls = getModelsArray('recalls', body);
            expect(recalls.length).toBe(1);
            expect(body).toMatchSnapshot();
          });
      });

      test('should return 403 for different accountIds', () => {
        return request(app)
          .post(`${rootUrl}/${accountId2}/recalls`)
          .set('Authorization', `Bearer ${token}`)
          .expect(403);
      });
    });

    describe('PUT /:accountId/recalls/:recallId', () => {
      test('should update a recall for the account', () => {
        return request(app)
          .put(`${rootUrl}/${accountId}/recalls/${recallId1}`)
          .set('Authorization', `Bearer ${token}`)
          .send({
            id: newRecallId,
            primaryType: 'phone',
            createdAt: '2017-07-19T00:14:30.932Z',
          })
          .expect(200)
          .then(({ body }) => {
            const recalls = getModelsArray('recalls', body);
            const [recall] = recalls;
            expect(recalls.length).toBe(1);
            expect(recall.primaryType).toBe('phone');
            expect(body).toMatchSnapshot();
          });
      });

      test('should return 403 for different accountIds', () => {
        return request(app)
          .put(`${rootUrl}/${accountId2}/recalls/${recallId1}`)
          .set('Authorization', `Bearer ${token}`)
          .expect(403);
      });
    });

    describe('DELETE /:accountId/recalls/:recallId', () => {
      afterEach(async () => {
        // have to restore recalls cause these routes could delete
        await seedRecalls();
      });

      test('should delete a recall for the account', () => {
        return request(app)
          .delete(`${rootUrl}/${accountId}/recalls/${recallId1}`)
          .set('Authorization', `Bearer ${token}`)
          .expect(204);
      });

      test('should return 403 for different accountIds', () => {
        return request(app)
          .delete(`${rootUrl}/${accountId2}/recalls/${recallId1}`)
          .set('Authorization', `Bearer ${token}`)
          .expect(403);
      });
    });
  });

});
