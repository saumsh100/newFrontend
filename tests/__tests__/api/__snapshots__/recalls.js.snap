// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/accounts/:account/recalls Recalls GET /:accountId/recalls should fetch all recalls for the account 1`] = `
Object {
  "entities": Object {
    "recalls": Object {
      "d5ab9bc0-f0e6-4538-99ae-2fe7f920abf4": Object {
        "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
        "id": "d5ab9bc0-f0e6-4538-99ae-2fe7f920abf4",
        "interval": "{\\"-1 months\\"}",
        "isActive": true,
        "isDeleted": false,
        "lengthSeconds": null,
        "primaryType": "email",
        "primaryTypes": Array [
          "sms",
        ],
      },
      "e5ab9bc0-f0e6-4538-99ae-2fe7f920abf4": Object {
        "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
        "id": "e5ab9bc0-f0e6-4538-99ae-2fe7f920abf4",
        "interval": "{\\"1 months\\"}",
        "isActive": true,
        "isDeleted": false,
        "lengthSeconds": null,
        "primaryType": "email",
        "primaryTypes": Array [
          "sms",
        ],
      },
    },
  },
  "result": Array [
    "d5ab9bc0-f0e6-4538-99ae-2fe7f920abf4",
    "e5ab9bc0-f0e6-4538-99ae-2fe7f920abf4",
  ],
}
`;

exports[`/api/accounts/:account/recalls Recalls GET /:accountId/recalls/outbox Outbox for Recalls - Should return Patient Old 1`] = `
Array [
  Object {
    "patient": Object {
      "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
      "address": Object {},
      "avatarUrl": null,
      "birthDate": null,
      "contCareInterval": "6 months",
      "contactMethodNote": null,
      "dueForHygieneDate": null,
      "dueForRecallExamDate": null,
      "email": "hello@hello.com",
      "familyId": null,
      "firstApptDate": null,
      "firstApptId": null,
      "firstName": "Old",
      "gender": null,
      "homePhoneNumber": null,
      "hygiene": true,
      "insurance": null,
      "insuranceInterval": null,
      "isDeleted": false,
      "isSyncedWithPMS": false,
      "isSyncedWithPms": false,
      "language": null,
      "lastApptDate": null,
      "lastApptId": null,
      "lastHygieneApptId": null,
      "lastName": "Patient",
      "lastRecallApptId": null,
      "lastRecallDate": null,
      "lastRestorativeApptId": null,
      "lastRestorativeDate": null,
      "middleName": null,
      "mobilePhoneNumber": null,
      "nextApptDate": null,
      "nextApptId": null,
      "otherPhoneNumber": null,
      "patientUserId": null,
      "phoneNumber": null,
      "pmsCreatedAt": null,
      "pmsId": null,
      "prefContactPhone": null,
      "prefName": null,
      "preferences": Object {
        "afternoon": true,
        "birthdayMessage": true,
        "emailNotifications": true,
        "evening": true,
        "morning": true,
        "newsletter": true,
        "phone": true,
        "recalls": true,
        "referrals": true,
        "reminders": true,
        "reviews": true,
        "sms": true,
        "weekdays": true,
        "weekends": true,
      },
      "sentRecalls": Array [],
      "status": "Active",
      "type": null,
      "workPhoneNumber": null,
    },
    "primaryTypes": Array [
      "email",
    ],
    "recall": Object {
      "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
      "interval": "1 months",
      "isActive": true,
      "isDeleted": false,
      "lengthSeconds": null,
      "primaryType": "email",
      "primaryTypes": Array [
        "email",
      ],
    },
    "sendDate": "2017-01-06T01:00:00.000Z",
  },
]
`;

exports[`/api/accounts/:account/recalls Recalls POST /:accountId/recalls should create a recall for the account 1`] = `
Object {
  "entities": Object {
    "recalls": Object {
      "f5ab9bc0-f0e6-4538-99ae-2fe7f920abf4": Object {
        "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
        "id": "f5ab9bc0-f0e6-4538-99ae-2fe7f920abf4",
        "interval": "-2 months",
        "isActive": true,
        "isDeleted": false,
        "lengthSeconds": null,
        "primaryType": "email",
        "primaryTypes": Array [
          "sms",
        ],
      },
    },
  },
  "result": "f5ab9bc0-f0e6-4538-99ae-2fe7f920abf4",
}
`;

exports[`/api/accounts/:account/recalls Recalls PUT /:accountId/recalls/:recallId should update a recall for the account 1`] = `
Object {
  "entities": Object {
    "recalls": Object {
      "d5ab9bc0-f0e6-4538-99ae-2fe7f920abf4": Object {
        "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
        "id": "d5ab9bc0-f0e6-4538-99ae-2fe7f920abf4",
        "interval": "{\\"-1 months\\"}",
        "isActive": true,
        "isDeleted": false,
        "lengthSeconds": null,
        "primaryType": "email",
        "primaryTypes": Array [
          "phone",
        ],
      },
    },
  },
  "result": "d5ab9bc0-f0e6-4538-99ae-2fe7f920abf4",
}
`;
